### PR TITLE
Habana DLC Perf/TestSuite TF/PT tests -- gaudi test suite

### PR DESCRIPTION
**Pull request by omrialmog**
* Add Habana DLAMI Tensorflow Performance Benchmarks
* Add Habana DLAMI PyTorch Performance Benchmarks
* Add Habana DLAMI Tensorflow Test Suite
* Add Habana DLAMI PyTorch Test Suite

*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Review by SergTogul — DISMISSED** [review]

File: test/dlc_tests/benchmark/ec2/pytorch/training/test_performance_pytorch_training.py [hunk 1]
```diff
@@ -58,12 +67,53 @@ def test_performance_pytorch_gpu_imagenet(pytorch_training, ec2_connection, gpu_
...
 
-# Placeholder for habana benchmark test
-@pytest.mark.model('N/A')
-#@pytest.mark.parametrize("ec2_instance_type", [PT_EC2_HPU_INSTANCE_TYPE], indirect=True)
-#@pytest.mark.parametrize("ec2_instance_ami", [HPU_AL2_DLAMI], indirect=True)
-def test_performance_tensorflow_hpu_imagenet(pytorch_training_habana):
-    pass  
+@pytest.mark.model("resnet50")
+@pytest.mark.parametrize("ec2_instance_type", [PT_EC2_HPU_INSTANCE_TYPE], indirect=True)
+# TODO: Ensure 8 card instance is used
```
Comment (tejaschumbalkar) [inline test/dlc_tests/benchmark/ec2/pytorch/training/test_performance_pytorch_training.py:72]: nit: lets remove the comment

File: test/dlc_tests/benchmark/ec2/pytorch/training/test_performance_pytorch_training.py [hunk 2]
```diff
@@ -58,12 +67,53 @@ def test_performance_pytorch_gpu_imagenet(pytorch_training, ec2_connection, gpu_
...
+        post_process=post_process_pytorch_hpu_py3_synthetic_ec2_training_performance,
+        data_source="synthetic",
+        threshold={"Throughput": threshold},
+        cards_num=cards_num,
+        synapseai_version=synapseai_version
+    )
+
+@pytest.mark.model("bert")
+@pytest.mark.parametrize("ec2_instance_type", [PT_EC2_HPU_INSTANCE_TYPE], indirect=True)
+# TODO: Ensure 8 card instance is used
```
Comment (tejaschumbalkar) [inline test/dlc_tests/benchmark/ec2/pytorch/training/test_performance_pytorch_training.py:96]: nit: lets remove the comment

File: test/dlc_tests/benchmark/ec2/tensorflow/training/test_performance_tensorflow_training.py [hunk 3]
```diff
@@ -75,14 +88,89 @@ def test_performance_tensorflow_gpu_imagenet(tensorflow_training, ec2_connection
         threshold={"Throughput": threshold},
     )
 
+@pytest.mark.integration("synthetic dataset")
+@pytest.mark.model("resnet50")
+@pytest.mark.parametrize("ec2_instance_type", [TF_EC2_HPU_INSTANCE_TYPE], indirect=True)
+# TODO: Ensure 8 card instance is used
```
Comment (tejaschumbalkar) [inline test/dlc_tests/benchmark/ec2/tensorflow/training/test_performance_tensorflow_training.py:94]: nit: lets remove the comment

File: test/dlc_tests/benchmark/ec2/tensorflow/training/test_performance_tensorflow_training.py [hunk 4]
```diff
@@ -75,14 +88,89 @@ def test_performance_tensorflow_gpu_imagenet(tensorflow_training, ec2_connection
...
+        data_source="synthetic",
+        threshold={"Throughput": threshold},
+        cards_num=cards_num,
+        synapseai_version=synapseai_version
+    )
+
+@pytest.mark.integration("squad dataset")
+@pytest.mark.model("bert")
+@pytest.mark.parametrize("ec2_instance_type", [TF_EC2_HPU_INSTANCE_TYPE], indirect=True)
+# TODO: Ensure 8 card instance is used
```
Comment (tejaschumbalkar) [inline test/dlc_tests/benchmark/ec2/tensorflow/training/test_performance_tensorflow_training.py:119]: nit: lets remove the comment

File: test/dlc_tests/benchmark/ec2/pytorch/training/test_performance_pytorch_training.py (hunk 1 above)
Comment (omrialmog) [reply test/dlc_tests/benchmark/ec2/pytorch/training/test_performance_pytorch_training.py]: Done

File: test/dlc_tests/benchmark/ec2/pytorch/training/test_performance_pytorch_training.py (hunk 2 above)
Comment (omrialmog) [reply test/dlc_tests/benchmark/ec2/pytorch/training/test_performance_pytorch_training.py]: Done

File: test/dlc_tests/benchmark/ec2/tensorflow/training/test_performance_tensorflow_training.py (hunk 3 above)
Comment (omrialmog) [reply test/dlc_tests/benchmark/ec2/tensorflow/training/test_performance_tensorflow_training.py]: Done

File: test/dlc_tests/benchmark/ec2/tensorflow/training/test_performance_tensorflow_training.py (hunk 4 above)
Comment (omrialmog) [reply test/dlc_tests/benchmark/ec2/tensorflow/training/test_performance_tensorflow_training.py]: Done

File: test/dlc_tests/container_tests/bin/benchmark/run_pytorch_rn50_training_performance_hpu_synthetic [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+mkdir -p ${LOG_DIR}
+
+LINUX_DIST_NAME=`cat /etc/*-release | grep "^NAME=" | sed 's#^.*=##' | tr -d '"'`
+if [ -z "$LINUX_DIST_NAME" ]; then
+    echo Unable to identify Linux distribution
+    exit 1
+elif [ "$LINUX_DIST_NAME" == "Ubuntu" ]; then
+    apt-get update
+    apt-get install -y imagemagick
+elif [ "$LINUX_DIST_NAME" == "Amazon Linux" ]; then
```
Comment (shantanutrip) [inline test/dlc_tests/container_tests/bin/benchmark/run_pytorch_rn50_training_performance_hpu_synthetic:20]: Is this required? Are we running on AL?

**Review by shantanutrip — COMMENTED** [review]
LGTM. Soft approved.

File: test/dlc_tests/container_tests/bin/benchmark/run_pytorch_rn50_training_performance_hpu_synthetic (hunk 5 above)
Comment (omrialmog) [reply test/dlc_tests/container_tests/bin/benchmark/run_pytorch_rn50_training_performance_hpu_synthetic:20]: Not necessarily, but doesnt hurt to leave it there in case AML2 will be released. (Its matching DLAMI as well)

File: test/test_utils/ec2.py [hunk 6]
```diff
@@ -536,6 +544,35 @@ def execute_ec2_training_performance_test(
...
+    docker_cmd = "docker"
+    container_test_local_dir = os.path.join("$HOME", "container_tests")
+
+    timestamp = time.strftime("%Y-%m-%d-%H-%M-%S")
+    log_name = f"{data_source}_results_{os.getenv('CODEBUILD_RESOLVED_SOURCE_VERSION')}_{timestamp}.txt"
+    synapseai_version = get_synapseai_version_from_tag(ecr_uri)
+    # Make sure we are logged into ECR so we can pull the image
+    connection.run(f"$(aws ecr get-login --no-include-email --region {region})", hide=True)
+
+    connection.run(f"{docker_cmd} pull -q {ecr_uri}")
```
Comment (SergTogul) [inline test/test_utils/ec2.py:558]: Is there a special purpose to preload the container?

File: test/testrunner.py [hunk 7]
```diff
@@ -345,13 +345,22 @@ def main():
         # Execute dlc_tests pytest command
         pytest_cmd = ["-s", "-rA", test_path, f"--junitxml={report}", "-n=auto"]
 
+        is_habana_image = any("habana" in image_uri for image_uri in all_image_list)
         if specific_test_type == "ec2":
+            if is_habana_image:
+                context = Context()
+                context.run("git clone https://github.com/HabanaAI/gaudi-test-suite.git")
```
Comment (SergTogul) [inline test/testrunner.py:352]: Why is it here? It a prerequisite for habana test only, right? So it should be somewhere in ec2 test fixtures?

File: habana/pytorch/training/docker/1.10/py3/synapseai1.2.0/Dockerfile.hpu [hunk 8]
```diff
@@ -31,12 +31,12 @@ RUN apt-get update \
...
 
 
-ARG PT_TRAINING_URL=https://aws-habana-pytorch-cicd-v3-binaries.s3.us-west-2.amazonaws.com/r1.10.0_aws_v3_habana/20220119-001148/972afcbe0f5754fd4f41d087730bd2870967c606/al2/torch-1.10.0-cp38-cp38-manylinux1_x86_64.whl
+ARG PT_TRAINING_URL=https://aws-habana-pytorch-cicd-v3-binaries.s3.us-west-2.amazonaws.com/r1.10.0_aws_v3_habana_telemetry/torch-1.10.0-cp38-cp38-manylinux1_x86_64.whl
 ARG PT_TORCHVISION_URL=https://aws-habana-pytorch-cicd-v3-binaries.s3.us-west-2.amazonaws.com/r1.10.0_aws_v3_habana/torchvision/torchvision-0.11.1-cp38-cp38-linux_x86_64.whl
 
 RUN pip --no-cache-dir install --upgrade \
     pip \
-    setuptools
+    "setuptools==58.2.0"
```
[the remark this answers is not in the document]
Comment (tejaschumbalkar) [reply habana/pytorch/training/docker/1.10/py3/synapseai1.2.0/Dockerfile.hpu:39]: The dependency is mapped in habana base image with respect to the synapseai version v1.2

File: test/dlc_tests/container_tests/bin/benchmark/run_pytorch_bert_training_performance_hpu [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+cd ${HOME_DIR}/artifacts/gaudi-test-suite
+git checkout r${GIT_BRANCH}
+python3 -m pip install -r requirements.txt
+python3 -m pip install -r test/benchmark/pt-requirements.txt
+
+START=$(date +%s)
+TRAINING_TIME="200m"
+if [ "$PR_CONTEXT" == "1" ]
+then
+  TRAINING_TIME="150m"
```
[the remark this answers is not in the document]
Comment (tejaschumbalkar) [reply test/dlc_tests/container_tests/bin/benchmark/run_pytorch_bert_training_performance_hpu:23]: For now, yes. Habana team is going to work if we can optimize the time. Also, we are going to select tests to run on PR and mainline context.

File: test/test_utils/ec2.py (hunk 6 above)
Comment (tejaschumbalkar) [reply test/test_utils/ec2.py:558]: No, nothing special. The function definition is referred from `execute_ec2_training_performance_test`. Optionally we can skip the pull. Will get this addressed in separate PR

File: test/testrunner.py (hunk 7 above)
Comment (tejaschumbalkar) [reply test/testrunner.py:352]: Tried the `Session` or `autouse` fixtures, they seem to rerun for each function. I will deep dive on the pytest configuration and follow up on this.

**Review by SergTogul — DISMISSED** [review]

**Comment by tejaschumbalkar** [comment]
TF framework and bert 1 card test fails only on CI. Will follow up on the fix on a separate PR.
